theano/tensor/slinalg.py

```diff
@@ -413,7 +413,12 @@ def perform(self, node, inputs, outputs):
         U = scipy.linalg.inv(V).T
 
         exp_w = numpy.exp(w)
-        X = numpy.subtract.outer(exp_w, exp_w) / numpy.subtract.outer(w, w)
+        exp_diff = numpy.subtract.outer(exp_w, exp_w)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", RuntimeWarning)
+            X = exp_diff / numpy.subtract.outer(w, w)
+        X[exp_diff == 0] = 1
```

Comment:
I am not a Theano user and stumbled on this when looking for an implementation of the derivative of the exponential.

The division by zero arises in the computation of
```
(np.exp(x)-np.exp(y))/(x-y)
```
When `x` is very close to `y` we can use the Taylor expansion, e.g.
```
delta = x-y
np.exp(x)*(1 + delta/2 + delta*delta/6)
```
This removes the need for imposing `X[exp_diff == 0] = 1`

Overall, my current changes look like this
```
w, V = scipy.linalg.eig(A, right=True)
U = scipy.linalg.inv(V).T

def myloewner(x,y):
    delta = x-y
    if( abs(delta) < 1e-6 ):
        return np.exp(x)*(1 + delta/2 + delta*delta/6)
    return (np.exp(x)-np.exp(y))/(x-y)

myufunc = np.frompyfunc(myloewner, 2, 1)
ratio = myufunc.outer(w,w).astype(w.dtype)
Y = np.real( U.dot( np.multiply( V.T.dot(gA).dot(U), ratio ) ).dot(V.T) )
```

That being said, it looks like this function is just replicating the functionalities of [`scipy.linalg.expm_frechet`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.expm_frechet.html). Is this intentional to improve the speed or accuracy? Note that `scipy.linalg.expm_frechet` should also work for matrices that are not diagonalisable.

The reason I investigated this piece of code was because I wanted to have the complete derivative and not only the Fréchet derivative. This code allows to compute it fairly easily by using
```
np.kron(V, U).dot( np.diag(ratio.ravel()) ).dot( np.kron(U.T,V.T) ) 
```
but again this will only support matrices that are diagonalisable. If anyone has a good implementation of the complete derivative, I'd be quite interested. The following paper is a nice reference

> Najfeld and T. F. Havel, “Derivatives of the matrix exponential and their computation,” Advances in Applied Mathematics, vol. 16, no. 3, 1995. [Online]. Available: http: //dx.doi.org/10.1006/aama.1995.1017

but my implementation of it doesn't seem to be working very well (copied here for archiving purposes)
```
expA = np.expm(A)
IkeA = np.kron(np.eye(3), expA)
adA = np.kron(np.eye(3), A) - np.kron(A.T, np.eye(3))
funadA = scipy.linalg.funm(adA, lambda x: (1-np.exp(-x))/x )
DexpA = np.dot(IkeA, funadA)
```